### ISSUE TITLE
Optional transpile

### ISSUE THREAD

**Issue by tjstavenger-pnnl** [Qiskit/qiskit-machine-learning#567]
### What should we add?

Currently Qiskit Machine-learning simulates with a transpile step. If the system simulated has no specific device, I believe transpiling is optional. Could we add a flag to the `fit()` function (or other functions that eventually transpile) to make the transpilation step optional?

**Comment by adekusar-drl** [comment]
I believe it is possible to do and available options depend on what algorithm you run. 

If it is an old `QuantumInstance`-based algorithm like `QuantumKernel` or `CircuitQNN`/`OpflowQNN`, then you make take a look at a list of parameters available here: https://qiskit.org/documentation/stubs/qiskit.utils.QuantumInstance.html. For instance, you may set `optimization_level=0`.

If it is a new primitive-based algorithm like `FidelityQuantumKernel` or `SamplerQNN`/`EstimatorQNN`, then you may configure the underlying primitive whether it is `Sampler` or `Estimator` properly. The reference implementation, https://qiskit.org/documentation/apidoc/primitives.html, does not invoke transpiler at all, while the runtime primitives expose a lot of option that can be tuned: https://qiskit.org/documentation/partners/qiskit_ibm_runtime/apidocs/options.html. The runtime primitives are used to execute circuits in the cloud, on hardware or on a simulator.

Let me know if you have questions.

**Comment by tjstavenger-pnnl** [comment]
I tried:

```
        # options = qiskit_ibm_runtime.options.Options()
        # options.transpilation.skip_transpilation = True
        options = {
            "transpilation": {
                "skip_transpilation": True
            }
        }
        estimator = qiskit.primitives.Estimator(options = options)
        regression_estimator_qnn = EstimatorQNN(
            circuit=qc, 
            input_params=feature_map.parameters, 
            weight_params=ansatz.parameters,
            estimator=estimator
        )
```

When using the Options class, I get an error `TypeError: update_options() argument after ** must be a mapping, not Options`. When using the dictionary, the transpilation still happens. Presumably I'm using the Options class wrong and I've setup the dictionary incorrectly?

We're running the tutorial here https://qiskit.org/documentation/machine-learning/tutorials/02_neural_network_classifier_and_regressor.html#Regression-with-an-EstimatorQNN but with custom gates that require the parameters to be bound before transpilation. The `Parameter(x)` is bound for some of the shots during the simulation, but eventually is unbound and fails. The failure came during the transpilation step, so I thought maybe if we were to skip the transpile (as it isn't needed in our case), we might be OK.

**Comment by tjstavenger-pnnl** [comment]
A part of what led me to ask for a way to turn off transpilation is that for the call stack I get when running the previously linked tutorial, there didn't appear to be a check to see if transpilation was necessary or not -- it was just called. Here's the full call stack: 

MainThread:
```
_wait_for_tstate_lock (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\threading.py:1027)
join (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\threading.py:1011)
shutdown (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\concurrent\futures\thread.py:236)
__exit__ (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\concurrent\futures\_base.py:636)
submit (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\primitives\primitive_job.py:44)
run (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\algorithms\gradients\base_estimator_gradient.py:98)
_backward (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit_machine_learning\neural_networks\estimator_qnn.py:245)
backward (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit_machine_learning\neural_networks\neural_network.py:252)
gradient (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit_machine_learning\algorithms\objective_functions.py:129)
wrapped_gradient (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\algorithms\optimizers\scipy_optimizer.py:172)
grad_wrapped (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\scipy\optimize\_differentiable_functions.py:164)
update_grad (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\scipy\optimize\_differentiable_functions.py:167)
_update_grad (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\scipy\optimize\_differentiable_functions.py:256)
__init__ (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\scipy\optimize\_differentiable_functions.py:177)
_prepare_scalar_function (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\scipy\optimize\_optimize.py:263)
_minimize_lbfgsb (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\scipy\optimize\_lbfgsb_py.py:306)
minimize (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\scipy\optimize\_minimize.py:699)
minimize (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\algorithms\optimizers\scipy_optimizer.py:148)
_fit_internal (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit_machine_learning\algorithms\regressors\neural_network_regressor.py:46)
fit (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit_machine_learning\algorithms\trainable_model.py:201)
test_bosnic_qiskit (c:\Users\stav405\git\bosonic-qiskit\tests\test_ml.py:104)
pytest_pyfunc_call (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\python.py:183)
_multicall (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\callers.py:187)
<lambda> (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:84)
_hookexec (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:93)
__call__ (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\hooks.py:286)
runtest (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\python.py:1641)
pytest_runtest_call (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\runner.py:162)
_multicall (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\callers.py:187)
<lambda> (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:84)
_hookexec (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:93)
__call__ (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\hooks.py:286)
<lambda> (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\runner.py:255)
from_call (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\runner.py:311)
call_runtest_hook (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\runner.py:254)
call_and_report (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\runner.py:215)
runtestprotocol (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\runner.py:126)
pytest_runtest_protocol (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\runner.py:109)
_multicall (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\callers.py:187)
<lambda> (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:84)
_hookexec (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:93)
__call__ (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\hooks.py:286)
pytest_runtestloop (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\main.py:348)
_multicall (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\callers.py:187)
<lambda> (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:84)
_hookexec (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:93)
__call__ (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\hooks.py:286)
_main (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\main.py:323)
wrap_session (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\main.py:269)
pytest_cmdline_main (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\main.py:316)
_multicall (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\callers.py:187)
<lambda> (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:84)
_hookexec (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\manager.py:93)
__call__ (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\pluggy\hooks.py:286)
main (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\_pytest\config\__init__.py:162)
_run_code (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\runpy.py:87)
_run_module_as_main (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\runpy.py:194)
```

ThreadPoolExecutor:
```
_define (c:\Users\stav405\git\bosonic-qiskit\c2qa\operators.py:78)
definition (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\circuit\instruction.py:237)
run (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\transpiler\passes\basis\unroll_custom_definitions.py:77)
_run_this_pass (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\transpiler\runningpassmanager.py:201)
_do_pass (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\transpiler\runningpassmanager.py:172)
run (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\transpiler\runningpassmanager.py:125)
_run_single_circuit (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\transpiler\passmanager.py:283)
run (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\transpiler\passmanager.py:228)
run (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\transpiler\passmanager.py:528)
_serial_transpile_circuit (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\compiler\transpiler.py:475)
transpile (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\compiler\transpiler.py:382)
_make_param_shift_gradient_circuit_data (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\algorithms\gradients\utils.py:99)
_param_shift_preprocessing (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\algorithms\gradients\utils.py:209)
_run (c:\Users\stav405\git\bosonic-qiskit\venv\Lib\site-packages\qiskit\algorithms\gradients\param_shift_estimator_gradient.py:71)
run (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\concurrent\futures\thread.py:57)
_worker (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\concurrent\futures\thread.py:80)
run (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\threading.py:870)
_bootstrap_inner (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\threading.py:932)
_bootstrap (c:\Users\stav405\AppData\Local\Programs\Python\Python38\Lib\threading.py:890)
```

I found that where `transpile()` is called, and for the call stack above it, there wasn't a check for if it would need to run or not. I've done another search for the `skip_transpilation` option, and I don't see it used here, either. This is the code from the ThreadPoolExecutor call stack where `transpile()` is called: https://github.com/Qiskit/qiskit-terra/blob/stable/0.22/qiskit/algorithms/gradients/utils.py#L99

It looks like qiskit-terra has been updated since v0.22.0 that we're using. I could try updating to see if that helps?

**Comment by adekusar-drl** [comment]
> I tried:
> 
> ```
>         # options = qiskit_ibm_runtime.options.Options()
>         # options.transpilation.skip_transpilation = True
>         options = {
>             "transpilation": {
>                 "skip_transpilation": True
>             }
>         }
>         estimator = qiskit.primitives.Estimator(options = options)
>         regression_estimator_qnn = EstimatorQNN(
>             circuit=qc, 
>             input_params=feature_map.parameters, 
>             weight_params=ansatz.parameters,
>             estimator=estimator
>         )
> ```

You are mixing up `qiskit` primitives and options from `qiskit_ibm_runtime`. Nothing unexpected the code does not work. There are two sets of the primitives. One is the reference implementation and the second one is in the runtime. The latter is used to run algorithms in the cloud.

**Comment by adekusar-drl** [comment]
> I found that where transpile() is called, and for the call stack above it, there wasn't a check for if it would need to run or not. I've done another search for the skip_transpilation option, and I don't see it used here, either. This is the code from the ThreadPoolExecutor call stack where transpile() is called: https://github.com/Qiskit/qiskit-terra/blob/stable/0.22/qiskit/algorithms/gradients/utils.py#L99

Ah, I see. It is called in gradients. Basically, it is required for gradients. Well, optimization level is set to zero, so transpilation should be fast. Transpilation is required to convert a circuit to a circuit that contains only supported gates. For example, you may create a circuit, convert it to the instruction and add this instruction to another circuit. When you print/draw the second circuit you'll realize that the instruction(the first circuit) is represented as large box. By default, gradients don't know how to deal with such instructions and they break them down into smaller pieces by calling `transpile()`. Hope it makes sense.

**Comment by adekusar-drl** [comment]
> Though I'm still getting an unbound parameter during the fit for the same Qiskit Machine Learning tutorial I linked earlier. Perhaps a question for a different ticket or a different venue -- where in Qiskit Machine Learning is it binding the parameters? Are there conditions around when it will and won't bind them?

I don't fully understand the problem. You have to have as many inputs to a QNN as it defines and as many weights as it defines as well. If you could share a small snippet it would be easier to say what's wrong.

Primitive-based QNNs don't bind parameters at all, they create an array of values to be bound and pass it to a primitive instance. These values are bound by the primitive.

**Comment by tjstavenger-pnnl** [comment]
Thank you for continuing to help me understand where I'm going wrong and how the Qiskit Machine Learning package is supposed to work. I did find on my own as well that the transpilation is converting our custom gates into the basis gates, and subsequently calling our gates' `_define()` function where the parameterized values haven't been bound yet. We're representing our gates as Qiskit UnitaryGate, which needs the full operator matrix to be defined. So far I've found it needs to have the parameters bound to perform the calculations to build the matrix.

I have seen arrays that I assumed to be the parameter values, so I've been a bit confused why the `ParamterExpression` themselves not being bound. I think I'm still missing understanding that better.

A test caes of the earlier Qiskit Machine Learning tutorial I mentioned, but using our bosonic-qiskit custom gates instead: https://github.com/C2QA/bosonic-qiskit/blob/parameterized-ml/tests/test_ml.py#L22-L97

You'll see the unbound parameter failure in our `_define()` function when Github runs our tests: https://github.com/C2QA/bosonic-qiskit/actions/runs/4366708881/jobs/7637075367#step:6:2190

Our bosonic-qiskit repo is open for public viewing, but you may need to be signed in to Github to view the test case result.

**Comment by tjstavenger-pnnl** [comment]
I'm not sure this helps, but I put a counter on the calls into `__array__()` and found it was consistently called 40 times with bound parameters. The 41st call is unbound for `Parameter(x)`, which I believe would point to the `feature_map` circuit and `input_params` in the `EstimatorQNN`: https://github.com/C2QA/bosonic-qiskit/blob/parameterized-ml/tests/test_ml.py#L45

**Comment by adekusar-drl** [comment]
Looks like you have custom gates and you try use them in the QML's regressor. Honestly, I'm not an expert in how Qiskit gates should be implemented. So here I can't help much. You can either open an issue in the Qiskit Terra repository or ask on Slack/StackOverflow. What I'm pretty sure is that QML models combines input parameters given by the dataset with weights into a single array and pass this array to primitives. And then the underlying primitive binds these values. So, the QNN circuit must have `num_inputs + num_weights` parameters in total. If the number of values passed is smaller than the number of parameters, then one of the parameters will be unbound.

**Comment by tjstavenger-pnnl** [comment]
I do agree that verifying our custom parameterized UnitaryGate is implemented as needed by Qiskit will be good. It's worked for us up to the point of using QML, but that may not mean it is truly correct. I'll follow up with qiskit-terra, Slack, StackOverflow and people from elsehwere on IBM's Qiskit team to verify.

At the point of calling the regressor's `fit()` function, we pass 20 inputs & 20 weights, which makes my 40 count of runs with bound parameters make sense? But why the 41st run with unbound parameters? I may be missunderstanding what you're referring to.

**Comment by adekusar-drl** [comment]
You are right. If you have 20 inputs and 20 weights then your circuit should have 40 parameters. And 40 values must be bound to execute the circuit. Try to double check what `qc.num_parameters` returns, where `qc` is the circuit you pass to the regressor.

Just in case, here inputs and weights are combined and then they are passed to a primitive: https://github.com/Qiskit/qiskit-machine-learning/blob/main/qiskit_machine_learning/neural_networks/neural_network.py#L147

**Comment by tjstavenger-pnnl** [comment]
Interestingly when I run the non-bosonic version of the tutorial (all we really did was change out the RY gate for a bosonic R gate), the RY gate's `_define()` function is called 560 times for just the feature map / input params. Of those I've checked, they appear to all be bound. That's certainly greater than the 40 we were expecting, where the 41st is coming in unbound when using the Bosonic R gate. 

I'll keep working with the Qiskit support areas (Slack/StackOverflow) to see if I can confirm our custom gates are parameterized correctly. Though it may turn out to be a case where the two systems -- QML and Bosonic Qiskit -- aren't compatible.

**Comment by adekusar-drl** [comment]
Thanks for keeping me in the loop!

**Comment by adekusar-drl** [comment]
@tjstavenger-pnnl How is it going on? Should I keep the issue or close it?

**Comment by tjstavenger-pnnl** [comment]
Not much progress since we last spoke. The bosonic-qiskit code that uses parmaterized circuits works around an issue where Qiskit doesn't natively support parameterizing unitary gates. I think the way I implemented that workaround is a contributing factor. I've reached out to the original user that raised the issue in bosonic-qiskit, but haven't heard back yet. I last commented on their ticket last week, so it hasn't been too long.

I think we're safe closing this ticket. If I have more questions, I'll reach out again.
